### ISSUE TITLE
Integration with fzf's CTRL+R and up

### ISSUE THREAD

**Issue by vitalybe** [larkery/zsh-histdb#25]
I would really like it it had the following features:

* Integration with fzf history search (CTRL+R)
* Ability to press "up" to go back in history from histdb and maybe only from this session.

**Comment by vitalybe** [comment]
Regarding CTRL+R, here is what I got:

```
histdb-fzf-widget() {
  local selected num
  setopt localoptions noglobsubst noposixbuiltins pipefail 2> /dev/null
  selected=( $(histdb --sep 999 | awk -F'999' '{print $4}' |
    FZF_DEFAULT_OPTS="--height ${FZF_TMUX_HEIGHT:-40%} $FZF_DEFAULT_OPTS -n2..,.. --tiebreak=index --bind=ctrl-r:toggle-sort $FZF_CTRL_R_OPTS --query=${(qqq)LBUFFER} +m" $(__fzfcmd)) )

  LBUFFER=$selected
  zle redisplay
  typeset -f zle-line-init >/dev/null && zle zle-line-init
  
  return $ret
}

zle     -N   histdb-fzf-widget
bindkey '^R' histdb-fzf-widget
```

I am a complete newbie with zsh, so not sure if it is correct, but it works.

Thoughts?

**Comment by norpol** [comment]
Thank you so much! I've basically had no ctrl+r search since September, because I've switched to `histdb`. The autocompletion felt too scary, so I basically ended up not using any history (except manually searching via. histdb) since then.

[fzf-zsh](https://github.com/Wyntau/fzf-zsh) gives a good guide to setup [fzf](https://github.com/junegunn/fzf) on oh-my-zsh.

I just need to figure out how to come back to the old ctrl+r regex search via. fzf or histdb - but this solution already cures so much pain!

**Comment by vitalybe** [comment]
@norpol You're welcome! But... You can still use the regular zsh's ctrl+R, don't you?

**Comment by norpol** [comment]
No, since all my history is now stored in `histdb` ctrl+R isn't working anymore.
I haven't investigated how to confgure standard zsh ctrl-r to search in histdb, fzf looks from usability perspective nice. :-)
Also one short notice: Your snippet shows the history out of order at least for me.

**Comment by larkery** [comment]
I did have a go at implementing a replacment control-R widget using zle, but I didn't get something I liked enough to use. I couldn't get it to look as nice as normal control-R, which seems to be a special thing within zsh's C internals.

I will see if I can dig it out in case anyone wants to do more on it.

Personally I just use control-R and zsh's normal history and then look in histdb when the command is from another machine or is hard to find or what have you, but better integration would be nice.

**Comment by larkery** [comment]
Here we are: https://github.com/larkery/zsh/blob/e26d355479e09d0bedd2dd262cdad33c701a3e0f/sqlite-history.zsh

This is from an older version so probably not directly compatible, but you can see something about how to use this with ZLE to search on $BUFFER and stick the match into POSTDISPLAY. What I couldn't work out how to do is to style the matching parts like how reverse isearch does (underlining the match), and I really like that so I gave up there.

**Comment by larkery** [comment]
I had another look at this and will see if I can get it working. So far I have:

```
typeset -g HISTDB_ISEARCH_INPUT
typeset -g HISTDB_ISEARCH_N
typeset -g HISTDB_ISEARCH_MATCH

# TODO Underlining the match string
# TODO Fill out the keymap properly
# TODO Show more info about match (n, date, pwd, host)
# TODO Accept match and exit
# TODO Keys to switch to dir
# TODO Keys to limit match?

# make a keymap for histdb isearch
bindkey -N histdb-isearch main

_histdb_isearch_query () {
    if [[ -z $HISTDB_ISEARCH_INPUT ]]; then
       HISTDB_ISEARCH_MATCH=""
       return
    fi
    local query="select commands.argv from history left join commands
on history.command_id = commands.rowid
where commands.argv like '%$(sql_escape ${HISTDB_ISEARCH_INPUT})%'
order by history.rowid desc
limit 1
offset ${HISTDB_ISEARCH_N}"
    HISTDB_ISEARCH_MATCH=$(_histdb_query "$query")
}

_histdb_isearch_display () {
    BUFFER="${HISTDB_ISEARCH_MATCH}"
    POSTDISPLAY="
histdb: ${HISTDB_ISEARCH_INPUT} "
}

# define a self-insert command for it (requires other code)
self-insert-histdb-isearch () {
    HISTDB_ISEARCH_INPUT="${HISTDB_ISEARCH_INPUT}${KEYS}"
    _histdb_isearch_query
    _histdb_isearch_display
}

zle -N self-insert-histdb-isearch

_histdb-isearch () {
    HISTDB_ISEARCH_INPUT="${BUFFER}"
    HISTDB_ISEARCH_N=0
    BUFFER=""
    zle -K histdb-isearch
    _histdb_isearch_query
    _histdb_isearch_display
}

zle -N _histdb-isearch

bindkey '' _histdb-isearch
```

This depends on https://github.com/larkery/zsh/blob/master/self-insert-override.zsh which unfortunately is something partially or wholly implemented in loads of other zsh plugins, so it probably won't cooperate with them.

**Comment by larkery** [comment]
I just pushed a commit which adds a mostly-working reverse isearch feature. It's good enough for me to use it so it will probably get more work when I find something about it annoying.

**Comment by norpol** [comment]
Hi, thank you very much! Looks really promising, I really like the way metadata (time, path) are displayed in search. I've tested it last night and also updated my branch (see #20).


I'm a little bit confused about the keys I should press to get the `isearch` visible, I've added this for now:

`bindkey '^R' _histdb-isearch`

Also I need to figure out how to get the regex search feature enabled.
With `bindkey '^R' history-incremental-pattern-search-backward`
I could do for example something like:
`^Rssh*hostname*-R`
and easily get to some ssh forward I've used a while ago.

Again, thank you so much for your project and I hope I'll be able to contribute at some point :+1: .

**Comment by larkery** [comment]
The search goes straight into sqlite, so you can use sql wildcards as patterns. Your example would be `ssh%hostname%-R`. Unfortunately the match underline will hit the wrong bit of the string - I'm not totally sure how to get it right. I might be able to get sqlite to tell me that as well, or repeat the match in zsh.

**Comment by larkery** [comment]
Turns out that was easier to fix than I thought. As of now, * and ? work. However, % and _ in the input will confuse, as they are not escaped, I'll do that later.

**Comment by noyez** [comment]
> Also one short notice: Your snippet shows the history out of order at least for me.

I had this problem too, i just added `--tac` to the `FZF_DEFAULT_OPTS` variable in @vitalybe's script above.

**Comment by jdsutherland** [comment]
I didn't like that the list contained duplicates (same cmd in diff dirs), so I made the following change to remove them:
`histdb --sep 999 | awk -F'999' '!seen[$4]++' | awk -F'999' '{print $4}'`

I also added an fzf bind to remove an entry using `--forget`:
`ctrl-d:execute(source ~/.zinit/plugins/larkery---zsh-histdb/sqlite-history.zsh && histdb --forget {}`
